driver_test.go

```diff
@@ -12,26 +12,263 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package spannerdriver_test
+package spannerdriver
 
 import (
+	"cloud.google.com/go/spanner"
+	"context"
 	"database/sql"
+	"fmt"
 	"log"
+	"os"
+	"reflect"
+	"runtime/debug"
+	"strings"
+	"testing"
 
-	spannerdriver "github.com/rakyll/go-sql-driver-spanner"
+	// API/lib packages not imported by driver.
+	adminapi "cloud.google.com/go/spanner/admin/database/apiv1"
 	"google.golang.org/api/option"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+	"google.golang.org/grpc"
 )
 
-func ExampleDriver() {
-	driver := &spannerdriver.Driver{
-		Options: []option.ClientOption{
-			option.WithCredentialsFile("/path/to/service-account-key.json"),
-		},
+var (
+	dsn string
+)
+
+type Connector struct {
+	ctx         context.Context
+	client      *spanner.Client
+	adminClient *adminapi.DatabaseAdminClient
+}
+
+func NewConnector() (*Connector, error) {
+
+	// Configure emulator if set.
+	spannerHost, ok := os.LookupEnv("SPANNER_EMULATOR_HOST")
+
+	ctx := context.Background()
+
+	var adminClient *adminapi.DatabaseAdminClient
+	var err error
+
+	if ok {
+		adminClient, err = adminapi.NewDatabaseAdminClient(
+			ctx,
+			option.WithoutAuthentication(),
+			option.WithEndpoint(spannerHost),
+			option.WithGRPCDialOption(grpc.WithInsecure()))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		adminClient, err = adminapi.NewDatabaseAdminClient(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	dataClient, err := spanner.NewClient(ctx, dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	curs := &Connector{
+		ctx:         ctx,
+		client:      dataClient,
+		adminClient: adminClient,
+	}
+	return curs, nil
+}
+
+func (c *Connector) Close() {
+	c.client.Close()
+	c.adminClient.Close()
+}
+
+func init() {
+
+	var projectId, instanceId, databaseId string
+	var ok bool
+
+	// Get environment variables or set to default.
+	if instanceId, ok = os.LookupEnv("SPANNER_TEST_INSTANCE"); !ok {
+		instanceId = "test-instance"
+	}
+	if projectId, ok = os.LookupEnv("SPANNER_TEST_PROJECT"); !ok {
+		projectId = "test-project"
+	}
+	if databaseId, ok = os.LookupEnv("SPANNER_TEST_DBID"); !ok {
+		databaseId = "gotest"
+	}
+
+	// Derive data source name.
+	dsn = "projects/" + projectId + "/instances/" + instanceId + "/databases/" + databaseId
+}
+
+// Used to check if error contains expected string
+// If want is the empty string, no error is expected
+func ErrorContainsStr(err error, want string) bool {
+	if want == "" && err != nil {
+		return false
+	}
+	if err == nil {
+		return want == ""
+	}
+	return strings.Contains(err.Error(), want)
+}
+
+// Executes DDL statements.
+func executeDdlApi(curs *Connector, ddls []string) (err error) {
```

Comment:
This function and the one below have a named return however the functions are not using this as intended. Please remove the named return i.e.
```suggestion
func executeDdlApi(curs *Connector, ddls []string) (error) {
```

Reply:
fixed!